Improve XDR Disconnected Endpoints job playbook - take 2

ali-sawyer opened the pull request:
<!-- REMINDER: THIS IS A PUBLIC REPO DO NOT POST HERE SECRETS/SENSITIVE DATA -->
## Contributing to Cortex XSOAR Content
Make sure to register your contribution by filling the [contribution registration form](https://forms.gle/XDfxU4E61ZwEESSMA)

**The Pull Request will be reviewed only after the contribution registration form is filled.**

## Status
- [ ] In Progress
- [x] Ready
- [ ] In Hold - (Reason for hold)

## Related Issues
n/a

## Description
Follow-up from this PR: https://github.com/demisto/content/pull/36237

Make the following improvements to the XDR Disconnected Endpoints playbook:
1. Update integration to allow returning all endpoints from the `xdr-get-endpoints` command
2. Convert timestamps to human-readable timestamps for the report
3. Fix a typo
4. Gracefully handle the situation where no lost/disconnected endpoints are found

## Must have
- [ ] Tests
- [ ] Documentation

tcarmeli1 commented on Packs/CortexXDR/Integrations/CortexXDRIR/CortexXDRIR.yml:
```diff
@@ -3557,25 +3572,22 @@ script:
       - resolved_true_positive
     - description: Comment to append to updated alerts.
       name: comment
-      required: false
```
do not delete this one

tcarmeli1 commented on Packs/CortexXDR/Integrations/CortexXDRIR/CortexXDRIR.yml:
```diff
@@ -3557,25 +3572,22 @@ script:
     name: xdr-update-alert
   dockerimage: demisto/python3:3.11.9.107902
   isfetch: true
-  isfetch:xpanse: false
```
do not delete this one

tcarmeli1 commented on Packs/CortexXDR/Integrations/CortexXDRIR/CortexXDRIR.yml:
```diff
@@ -3557,25 +3572,22 @@ script:
   isremotesyncin: true
   isremotesyncout: true
   type: python
+  runonce: false
```
remove this one

tcarmeli1 commented on Packs/CortexXDR/Integrations/CortexXDRIR/CortexXDRIR.yml:
```diff
@@ -3557,25 +3572,22 @@ script:
 - Test XDR Playbook retrieve file command
-defaultmapperin: Cortex XDR - IR-mapper
-defaultmapperout: Cortex XDR - IR-out-mapper
 fromversion: 5.0.0
```
Please add unit test for the code you added under this file Packs/CortexXDR/Integrations/CortexXDRIR/CortexXDRIR_test.py.
[docs reference](https://xsoar.pan.dev/docs/integrations/unit-testing#write-your-unit-tests)

ali-sawyer replied:
@tcarmeli1 please clarify what unit tests are required. I didn't add any new functions or commands, I added a couple of arguments to a command and modified some existing functions/methods.

tcarmeli1 replied:
Since new arguments were added, unit tests should be implemented to ensure that the previous functionality remains intact and that the new functionality works as expected. you can see examples in the current testing file
Packs/CortexXDR/Integrations/CortexXDRIR/CortexXDRIR_test.py.

tomer-pan commented on Packs/CortexXDR/Playbooks/playbook-Cortex_XDR_disconnected_endpoints.yml:
```diff
@@ -543,27 +558,29 @@ view: |-
 - key: Email
-  value: {}
+  value:
+    simple: asawyer@paloaltonetworks.com
```
@ali-sawyer Please remove the hardcoded email

tomer-pan commented on Packs/CortexXDR/Playbooks/playbook-Cortex_XDR_disconnected_endpoints.yml:
```diff
@@ -543,27 +558,29 @@ view: |-
 - key: LastSeenStartDate
-  value: {}
+  value:
+    simple: 7 days
```
@ali-sawyer Same here - please remove hardcoded value

tomer-pan commented on Packs/CortexXDR/Playbooks/playbook-Cortex_XDR_disconnected_endpoints.yml:
```diff
@@ -543,27 +558,29 @@ view: |-
 - key: LastSeenEndDate
-  value: {}
+  value:
+    simple: 0 days
```
@ali-sawyer Same here - please remove hardcoded value

tomer-pan commented:
Hi @ali-sawyer As for the playbook review,
I added some comments about the playbook inputs hardcoded values.
Besides that - looks good, thanks for your contribution!

ali-sawyer commented:
@tcarmeli1 The SDK did not download the correct python code with my changes for some reason, fixed it! I addressed most of your other comments as well. Please let me know if you have other feedback now that you have the correct code to review.

ali-sawyer commented:
@tomer-pan Oops, good catches on the hardcoded values, fixed.

tcarmeli1 commented on Packs/CortexXDR/Integrations/CortexXDRIR/CortexXDRIR.py:
```diff
@@ -1,13 +1,4473 @@
+
+    # Convert Unix epoch timestamps to human-readable timestamps
+    if use_hr_timestamps:
+        for endpoint in endpoints:
```
lets move this to a function (from line 2188-2198)

tcarmeli1 commented on Packs/CortexXDR/Integrations/CortexXDRIR/CortexXDRIR.py:
```diff
@@ -1,13 +1,4473 @@
+    )
+
+    # Return all endpoints vs single page
+    if get_all:
```
- change the arg name to all_results
- lets move this to a function (2158-2184)
- if we reach to the end- an empty array is returned from the api? we do not receive an error right?

ali-sawyer replied:
@tcarmeli1 Done.

Re: bullet 3, exactly: the API will return an empty array and no error. This tells us we are done paging over all the endpoints so we can stop looping at that point.

tcarmeli1 replied:
Amazing thank!

tcarmeli1 commented on Packs/CortexXDR/Integrations/CortexXDRIR/CortexXDRIR.yml:
```diff
@@ -917,6 +919,20 @@ script:
+    - auto: PREDEFINED
+      defaultValue: "false"
+      description: Whether to return all endpoints. If true, will override the 'limit' and 'page' arguments.
+      name: get_all
```
change the arg name to all_results

tcarmeli1 commented on Packs/CortexXDR/Integrations/CortexXDRIR/CortexXDRIR.yml:
```diff
@@ -183,6 +181,10 @@ configuration:
   advanced: true
   type: 8
   section: Collect
+  display: Incidents Fetch Interval
```
remove this please

tcarmeli1 commented on Packs/CortexXDR/Integrations/CortexXDRIR/CortexXDRIR.py:
```diff
@@ -1,13 +1,4473 @@
+def get_endpoints_command(client, args):
+    integration_context_brand = args.pop('integration_context_brand', 'CoreApiModule')
+    integration_name = args.pop("integration_name", "CoreApiModule")
+    get_all = argToBoolean(args.get('get_all', False))
```
please change the name to all_results

tcarmeli1 reviewed (changes requested):
Hi @ali-sawyer ,
Great work :)
Thank you for your response!
Added some more comments
Please let me know if you need any help.
You can contact me via slack either, Tal Carmeli

tcarmeli1 commented on Packs/ApiModules/Scripts/CoreIRApiModule/CoreIRApiModule.py:
```diff
@@ -2059,20 +2059,74 @@ def generate_endpoint_by_contex_standard(endpoints, ip_as_string, integration_na
     return standard_endpoints
 
 
+def page_over_endpoints(client, endpoints, endpoint_id_list, dist_name, ip_list, public_ip_list,
```
please change to a more informative function name - eg. retrieve_all_endpoints

tcarmeli1 commented on Packs/ApiModules/Scripts/CoreIRApiModule/CoreIRApiModule.py:
```diff
@@ -2059,20 +2059,74 @@ def generate_endpoint_by_contex_standard(endpoints, ip_as_string, integration_na
-        required=True
-    )
+    all_results = argToBoolean(args.get('all_results', False))
+    # If we are getting all endpoints, start at page 0 and use the max limit supported by the API of 100
```
please change to a more informative- 
When we desire to get all endpoints, start at page 0 and use the max limit supported by the API of (100)

tcarmeli1 commented on Packs/ApiModules/Scripts/CoreIRApiModule/CoreIRApiModule.py:
```diff
@@ -2132,6 +2187,19 @@ def get_endpoints_command(client, args):
         username=username
     )
 
+    # Return all endpoints vs single page
```
please remove the comment

tcarmeli1 reviewed (changes requested):
Hi Ali,
Thank you for your response and updates.
Great work! :)
Please see my comments below.
In addition, please fix the conflicts.
Thank you again.

ShirleyDenkberg commented on Packs/CortexXDR/ReleaseNotes/6_1_82.md:
```diff
@@ -0,0 +1,12 @@
+
+##### Cortex XDR - IR
+
+- Added support for returning all endpoints and outputting human-readable timestamps to the command **xdr-get-endpoints**.
+
+#### Playbooks
+
+##### Cortex XDR disconnected endpoints
+
+- Improved implementation of the playbook by returning all endpoints, outputting data with human-readable timestamps, and gracefully handling the situation where no lost or disconnected endpoints are found.
```
```suggestion
Added support for returning all endpoints and outputting human-readable timestamps to the ***xdr-get-endpoints*** command.

#### Playbooks

##### Cortex XDR disconnected endpoints

Improved implementation of the playbook by returning all endpoints, outputting data with human-readable timestamps, and gracefully handling the situation where no lost or disconnected endpoints are found.
```

ShirleyDenkberg commented:
@tomer-pan @tcarmeli1 @maimorag Doc review completed.

tcarmeli1 reviewed (changes requested):
Hi @ali-sawyer ,
Thank you for your great contribution.
Amazing work!
Since we made a lot of changes, I would appreciate if you can make a new demo video of the comments and playbook after all the changes.
Thank you so much!

tcarmeli1 reviewed (approved):
Great work!
